pkg/addonfactory/defaultagentaddon.go

```diff
@@ -0,0 +1,128 @@
+package addonfactory
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/openshift/library-go/pkg/assets"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// Template config name in annotation
+const templateConfigName string = "templateConfig"
+
+// Template field name
+const (
+	ClusterName           string = "ClusterName"
+	AddonInstallNamespace string = "AddonInstallNamespace"
+	KubeConfigSecret      string = "KubeConfigSecret"
+)
+
+type DefaultAgentAddon struct {
+	decoder           runtime.Decoder
+	templateData      map[string][]byte
+	templateConfig    interface{}
+	agentAddonOptions agent.AgentAddonOptions
+}
+
+func newDefaultAgentAddon(
+	scheme *runtime.Scheme,
+	templateConfig interface{},
+	agentAddonOptions agent.AgentAddonOptions) *DefaultAgentAddon {
+	return &DefaultAgentAddon{
+		decoder:           serializer.NewCodecFactory(scheme).UniversalDeserializer(),
+		templateData:      map[string][]byte{},
+		templateConfig:    templateConfig,
+		agentAddonOptions: agentAddonOptions}
+}
+
+func (a *DefaultAgentAddon) Manifests(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	var objects []runtime.Object
+
+	manifestConfig, err := a.getManifestConfig(cluster, addon)
+	if err != nil {
+		return objects, err
+	}
+	for file, data := range a.templateData {
+		raw := assets.MustCreateAssetFromTemplate(file, data, manifestConfig).Data
+		object, _, err := a.decoder.Decode(raw, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+
+	return objects, nil
+}
+
+func (a *DefaultAgentAddon) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return a.agentAddonOptions
+}
+
+func (a *DefaultAgentAddon) getManifestConfig(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (map[string]interface{}, error) {
+	manifestConfig := map[string]interface{}{}
+	userConfig := a.templateConfig
+
+	typ := reflect.TypeOf(userConfig)
+	val := reflect.ValueOf(userConfig)
+	for i := 0; i < typ.NumField(); i++ {
+		tf := typ.Field(i)
+		vf := val.Field(i)
+		manifestConfig[tf.Name] = vf.Interface()
+	}
+
+	manifestConfig[ClusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	manifestConfig[AddonInstallNamespace] = installNamespace
+	manifestConfig[KubeConfigSecret] = fmt.Sprintf("%s-hub-kubeconfig", a.agentAddonOptions.AddonName)
+
+	// get templateConfig from annotation of addon cr
+	annotations := addon.GetAnnotations()
```

Comment:
why we need this? and this needs to be well documented.

Reply:
doced.

Reply:
let's remove this part, but let user to define something similar like what we do with values in helmagent